### PR TITLE
More flexible Python Callable type.

### PR DESCRIPTION
**Pull request by robertwb**
This allows `def` functions, fully qualified imports, and expressions with preambles. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @ihji

**Comment by ihji** [comment]
I didn't realize you wrote the similar PR I drafted last night: https://github.com/apache/beam/pull/17763/files
I'm okay with the suggested implementation in this PR but prefer the simpler solution that explicitly separates the statement source for `exec` and the expression source for `eval`. Using statements as an expression by our own semantics seems more confusing to me.

@chamikaramj Cham, Do you have any preference / idea?

**Comment by robertwb** [comment]
Oh, didn't realize you were looking at this as well. 

The primary advantage of this form is for anything that is difficult to express as a lambda, e.g. having to write

```
PythonCallableSource.of(
    "foo")
.withContext("def foo(...): ...")
```

feels a bit roundabout. I was also leaning towards making things like `PythonCallableSource.of("math.sqrt")`  or `PythonCallableSource.of("scipy.special.ellipj")` work with minimal overhead, and reduce the amount of overhead required for each calling SDK (go, typescript, ...) by keeping the representation a simple string.

**Comment by ihji** [comment]
@robertwb Okay. Then can you please add a detailed pydoc about the modified semantics we implemented for a string representation of Python codes? We might need users to know what statements and expressions are allowed and how to use them with examples.

**Comment by chamikaramj** [comment]
+1 for simple approach (single string) with a detailed pydoc. From an API perspective I like separating out the exec and eval parts but I agree with Roberts concern that this might feel unnatural for a Python user who just want to use a single Python script that needs some statements to build context.

**Comment by robertwb** [comment]
Added documentation.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/PythonCallableSource.java [hunk 1]
```diff
@@ -20,7 +20,27 @@
...
- * A wrapper object storing a Python code that can be evaluated to Python callables in Python SDK.
+ * A wrapper object storing a Python function definition that can be evaluated to Python callables
+ * in Python SDK.
+ *
+ * <p>The snippet of Python code can be a valid Python expression (such as {@code lambda x: x * x}
+ * or {str.upper}), a fully qualified name (such as {@code math.sin}), or a complete, multi-line
+ * function definition (such as {@code def foo(x): ...}).
+ *
+ * <p>Any lines preceding the function definition are first evaluated to provide context in which to
+ * define the function which can be useful to declare inports or any other needed values, e.g.
```
Comment (ihji) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PythonCallableSource.java:31]: imports

Comment (ihji) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PythonCallableSource.java:31]: imports

**Review by ihji — APPROVED** [review]
LGTM except one typo. Thanks!

**Pull request merged**